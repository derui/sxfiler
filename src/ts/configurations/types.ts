import { ObjectEnum } from "@/utils";

export type ItemKey<T> = readonly [string, string, string];
export type SectionKey = readonly [string, string];
export type CategoryKey = readonly [string];

export type Descriptor<T> = {
  readonly key: T;
  readonly displayName: string;
  readonly description: string;
};

export type CategoryDescriptor = Descriptor<CategoryKey>;

export type SectionDescriptor = Descriptor<SectionKey>;

export type ItemDescriptor = Descriptor<ItemKey<any>>;

export type SelectOption = {
  value: any;
  toString(): string;
  display: string;
};

export const ItemElementType = {
  TEXT: "TEXT",
  NUMBER: "NUMBER",
  BOOLEAN: "BOOLEAN",
  SELECT_ONE: "SELECT_ONE",
} as const;
export type ItemElementType = ObjectEnum<typeof ItemElementType>;

export type ItemKind<K extends ItemElementType> = {
  kind: K;
};

export type TextItemType = ItemKind<"TEXT"> & {
  defaultValue?: string;
};
export type NumberItemType = ItemKind<"NUMBER"> & {
  defaultValue?: number;
};
export type BooleanItemType = ItemKind<"BOOLEAN"> & {
  defaultValue: boolean;
};
export type SelectOneItemType = ItemKind<"SELECT_ONE"> & {
  options: SelectOption[];
  defaultValue: string;
};

export type ItemType = TextItemType | NumberItemType | BooleanItemType | SelectOneItemType;

export type Item = ItemDescriptor & {
  readonly type: ItemType;
};

export type Section = SectionDescriptor & {
  readonly items: Item[];
};

export type Category = CategoryDescriptor & {
  readonly sections: Section[];
};
