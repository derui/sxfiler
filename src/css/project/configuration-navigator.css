.configuration-navigator {
  display: block;

  /* background-color: rgb(var(--color-base02)); */
  background-color: var(--color-configuration-navigatorBackground);

  /* color: rgb(var(--color-base3)); */
  color: var(--color-configuration-navigatorText);
  box-shadow: var(--box-shadow);
}

.configuration-navigator__category {
  padding-bottom: var(--space-base);
  padding-left: var(--space-base);
  padding-right: var(--space-base);

  &:first-of-type {
    padding-top: var(--space-base);
  }
}

.configuration-navigator__category-name {
  &:hover {
    cursor: pointer;
  }

  &::before {
    display: inline-block;
    margin-right: var(--space-small);
    content: "";
    width: 0;
    height: 0;
    border-style: solid;
    border-width: 5px 0 5px 8.7px;

    /* border-color: transparent transparent transparent rgb(var(--color-base3)); */
    border-color: transparent transparent transparent var(--color-configuration-navigatorCategoryMarker);
    transition: transform ease-out 100ms;
  }
}

.configuration-navigator__category-name--opened {
  &::before {
    transform: rotate(0.25turn);
  }
}

.configuration-navigator__section-tree {
  padding: 0;
  margin: 0;
  list-style: none;
  margin-left: calc(var(--space-base) + var(--space-large));

  &[aria-hidden="true"] {
    display: none;
  }
}

.configuration-navigator__section {
  padding: 0;
  margin: 0;
  padding-top: var(--space-small);

  /* color: rgb(var(--color-base3)); */
  color: var(--color-configuration-navigatorSectionText);
  cursor: pointer;

  &:hover {
    /* color: rgb(var(--color-orange)); */
    color: var(--color-configuration-navigatorSectionHover);
  }
}
