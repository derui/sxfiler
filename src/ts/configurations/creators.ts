import { Section, Category, Item, ItemType, ItemKey, SectionDescriptor, CategoryDescriptor } from "./types";

export const createCategory = function (args: CategoryDescriptor & { sections: Section[] }): Category {
  return { ...args };
};

export const createSection = function (args: SectionDescriptor & { items: Item[] }): Section {
  return { ...args };
};

export const createItem = function (args: {
  key: ItemKey<any>;
  displayName: string;
  description: string;
  type: ItemType;
}): Item {
  const { key, ...rest } = args;
  return {
    key,
    displayName: rest.displayName,
    description: rest.description,
    type: rest.type,
  };
};
