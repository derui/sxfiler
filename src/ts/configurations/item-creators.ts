import { ItemElementType, ItemType, SelectOption } from "./types";

export const createText = (value?: string): ItemType => ({ kind: ItemElementType.TEXT, defaultValue: value });

export const createNumber = (value?: number): ItemType => ({ kind: ItemElementType.NUMBER, defaultValue: value });

export const createBoolean = (value: boolean): ItemType => ({ kind: ItemElementType.BOOLEAN, defaultValue: value });

export const createSelectOne = (options: SelectOption[], value: string): ItemType => ({
  kind: ItemElementType.SELECT_ONE,
  options,
  defaultValue: value,
});
