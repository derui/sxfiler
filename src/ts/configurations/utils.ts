import { ItemElementType, Item } from "./types";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedText = function (arg: Item, value: any): value is string {
  return arg.type.kind === ItemElementType.TEXT && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedNumber = function (arg: Item, value: any): value is number {
  return arg.type.kind === ItemElementType.NUMBER && value !== undefined && value !== null;
};

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export const isTypedBoolean = function (arg: Item, value: any): value is boolean {
  return arg.type.kind === ItemElementType.BOOLEAN && value !== undefined && value !== null;
};
