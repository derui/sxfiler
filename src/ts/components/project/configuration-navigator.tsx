// this component shows category tree with configuration structure.

import { h, FunctionalComponent } from "preact";
import classnames from "classnames";
import { Category, Section, CategoryKey } from "@/configurations/types";
import { useState, StateUpdater } from "preact/hooks";
import { qualified } from "@/configurations";

export type Props = {
  categories: Category[];
  onSectionSelected: (category: Category, section: Section) => void;
};

type CategoryTreeProps = {
  category: Category;
  opened: boolean;
  onToggle: (key: CategoryKey) => void;
  onSelected: (category: Category, section: Section) => void;
};

type SectionProps = {
  section: Section;
  selected: boolean;
  onSelected: (section: Section) => void;
};

const Section: FunctionalComponent<SectionProps> = ({ section, selected, onSelected }) => {
  const className = classnames({
    "configuration-navigator__section": true,
    "configuration-navigator__section--selected": selected,
  });
  return (
    <li class={className} data-testid="configuration-navigator-section" onClick={() => onSelected(section)}>
      {section.displayName}
    </li>
  );
};

const CategoryTree: FunctionalComponent<CategoryTreeProps> = ({ category, opened, onToggle, onSelected }) => {
  const key = category.key;
  const className = `
    configuration-navigator__category-name
    ${opened ? "configuration-navigator__category-name--opened" : ""}
  `;

  return (
    <div class="configuration-navigator__category" data-testid="configuration-navigator-category">
      <span class={className} data-testid="configuration-navigator-category-name" onClick={() => onToggle(key)}>
        {category.displayName}
      </span>
      <ul
        class="configuration-navigator__section-tree"
        data-testid="configuration-navigator-section-tree"
        aria-hidden={!opened}
      >
        {Object.entries(category.sections).map(([, value], index) => (
          <Section
            key={index}
            section={value}
            selected={false}
            onSelected={(section) => onSelected(category, section)}
          />
        ))}
      </ul>
    </div>
  );
};

const handleCategoryClick = (setState: StateUpdater<{ [p: string]: boolean }>) => (categoryKey: CategoryKey) => {
  setState((prev) => {
    const current = prev[qualified(categoryKey)] || false;
    return { ...prev, [qualified(categoryKey)]: !current };
  });
};

export const Component: FunctionalComponent<Props> = ({ categories, onSectionSelected }) => {
  const [state, setState] = useState({} as { [p: string]: boolean });

  return (
    <section class="configuration-navigator">
      {categories.map((v, index) => {
        const opened = state[qualified(v.key)] || false;
        return (
          <CategoryTree
            key={index}
            category={v}
            opened={opened}
            onSelected={onSectionSelected}
            onToggle={handleCategoryClick(setState)}
          />
        );
      })}
    </section>
  );
};
