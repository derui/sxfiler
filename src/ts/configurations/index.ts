import { SortType, SortTypeMap } from "@/generated/types_pb";
import { Category, Item, Section, ItemKey, SectionKey, CategoryKey } from "./types";
import { createCategory, createSection, createItem } from "./creators";
import * as ItemCreators from "./item-creators";

const itemKeyFrom = <T>(section: SectionKey, key: string): ItemKey<T> => [section[0], section[1], key];
const categoryKeyOf = (key: ItemKey<any>): CategoryKey => [key[0]];
const sectionKeyOf = (key: ItemKey<any>): SectionKey => [key[0], key[1]];
const sectionKeyFrom = (category: CategoryKey, key: string): SectionKey => [category[0], key];
export const qualified = (key: readonly string[]): string => key.join(".");

export const findItemBy = (key: ItemKey<any>, category: Category): Item | undefined => {
  const categoryKey = categoryKeyOf(key);
  const sectionKey = sectionKeyOf(key);
  if (qualified(category.key) !== qualified(categoryKey)) {
    return;
  }
  const section = category.sections.find((v) => v.key === sectionKey);
  if (!section) {
    return;
  }

  return section.items.find((v) => qualified(v.key) === qualified(key));
};

export const findSectionBy = (key: SectionKey, category: Category): Section | undefined => {
  return category.sections.find((v) => qualified(v.key) === qualified(key));
};

export const findValueBy = (configuration: { [p: string]: any }, item: Item): any => {
  const value = configuration[qualified(item.key)];
  if (value === undefined || value === null) {
    return;
  }

  return value;
};

export const categories = {
  general: ["general"],
} as const;

export const sections = {
  behaviors: sectionKeyFrom(categories.general, "behaviors"),
  theme: sectionKeyFrom(categories.general, "theme"),
} as const;

export const itemKeys = {
  general: {
    behaviors: {
      confirmationWhenDelete: itemKeyFrom<string>(sections.behaviors, "confirmation_when_delete"),
      maxNumberOfHistories: itemKeyFrom<number>(sections.behaviors, "max_number_of_histories"),
      defaultSortOrder: itemKeyFrom<SortTypeMap>(sections.behaviors, "default_sort_order"),
    },
    theme: {
      currentTheme: itemKeyFrom<string>(sections.theme, "current_theme"),
    },
  },
} as const;

export namespace Configurations {
  const general = createCategory({
    key: categories.general,
    displayName: "General",
    description: "General configurations",
    sections: [
      createSection({
        key: sections.behaviors,
        displayName: "Behaviors",
        description: "Behaviors",
        items: [
          createItem({
            key: itemKeys.general.behaviors.confirmationWhenDelete,
            displayName: "Confirmation when delete",
            description: "Need user confirmation when delete something",
            type: ItemCreators.createBoolean(true),
          }),
          createItem({
            key: itemKeys.general.behaviors.maxNumberOfHistories,
            displayName: "Maximum number of histories",
            description:
              "History is the history of moved directory in a pane. This setting defines maximum number of histories in a pane",
            type: ItemCreators.createNumber(100),
          }),
          createItem({
            key: itemKeys.general.behaviors.defaultSortOrder,
            displayName: "Default sort order of Pane",
            description: "This setting defines sort order of filer when open this application",
            type: ItemCreators.createSelectOne(
              [
                {
                  display: "Date",
                  toString() {
                    return `${SortType.DATE}`;
                  },
                  value: SortType.DATE,
                },
                {
                  display: "Name",
                  toString() {
                    return `${SortType.NAME}`;
                  },
                  value: SortType.NAME,
                },
                {
                  display: "Size",
                  toString() {
                    return `${SortType.SIZE}`;
                  },
                  value: SortType.SIZE,
                },
              ],
              `${SortType.NAME}`
            ),
          }),
        ],
      }),
      createSection({
        key: sections.theme,
        displayName: "Theme",
        description: "Theme",
        items: [
          createItem({
            key: itemKeys.general.theme.currentTheme,
            displayName: "Current theme",
            description: "Select current theme",
            type: ItemCreators.createText("default"),
          }),
        ],
      }),
    ],
  });

  export const definitions = {
    general,
  } as const;
}
