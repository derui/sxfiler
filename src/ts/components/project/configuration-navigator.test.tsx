import { h } from "preact";
import { render as renderAsText } from "preact-render-to-string";
import { Component } from "./configuration-navigator";
import { render, cleanup, fireEvent, act } from "@testing-library/preact";
import { createSection, createCategory } from "@/configurations/creators";

describe("Project", () => {
  describe("Configuration navigator", () => {
    afterEach(() => cleanup());

    test("show empty categories", () => {
      const tree = renderAsText(<Component categories={[]} onSectionSelected={() => {}} />);

      expect(tree).toMatchSnapshot();
    });

    test("render category and sections as tree", async () => {
      const category = createCategory({
        key: ["key"],
        displayName: "display",
        description: "desc",
        sections: [
          createSection({
            key: ["key", "key2"],
            displayName: "section",
            description: "desc",
            items: [],
          }),
        ],
      });
      const query = render(<Component categories={[category]} onSectionSelected={() => {}} />);

      const section = await query.findAllByTestId("configuration-navigator-section");

      expect(section).toHaveLength(1);
      expect(section[0]).toHaveTextContent("section");
    });

    test("open hidden sections when category is clicked", async () => {
      const category = createCategory({
        key: ["key"],
        displayName: "display",
        description: "desc",
        sections: [
          createSection({
            key: ["key", "key2"],
            displayName: "section",
            description: "desc",
            items: [],
          }),
        ],
      });
      const query = render(<Component categories={[category]} onSectionSelected={() => {}} />);

      const sectionTree = await query.findByTestId("configuration-navigator-section-tree");

      expect(sectionTree).toHaveAttribute("aria-hidden", "true");

      const categoryDom = await query.findByTestId("configuration-navigator-category-name");
      act(() => {
        fireEvent.click(categoryDom);
      });

      expect(sectionTree).toHaveAttribute("aria-hidden", "false");
    });

    test("callback function when seciton is clicked", async () => {
      const category = createCategory({
        key: ["key"],
        displayName: "display",
        description: "desc",
        sections: [
          createSection({
            key: ["key", "key2"],
            displayName: "section",
            description: "desc",
            items: [],
          }),
        ],
      });
      const callback = jest.fn();
      const query = render(<Component categories={[category]} onSectionSelected={callback} />);

      const categoryDom = await query.findByTestId("configuration-navigator-category-name");
      act(() => {
        fireEvent.click(categoryDom);
      });
      const sectionDom = await query.findByTestId("configuration-navigator-section");
      act(() => {
        fireEvent.click(sectionDom);
      });

      expect(callback).toBeCalledWith(category, category.sections[0]);
    });
  });
});
