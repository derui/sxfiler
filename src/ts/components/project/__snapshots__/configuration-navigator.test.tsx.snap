// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Configuration navigator show empty categories 1`] = `"<section class=\\"configuration-navigator\\"></section>"`;
